zcash_primitives/src/transaction/components/sapling/builder.rs

```diff
@@ -373,204 +462,401 @@ impl<P: consensus::Parameters> SaplingBuilder<P> {
-                                    break addr;
-                                }
-                            }
-                        };
-
-                        let rseed =
-                            generate_random_rseed_internal(&params, target_height, &mut rng);
-
-                        Note::from_parts(payment_address, NoteValue::from_raw(0), rseed)
-                    };
-
-                    let esk = dummy_note.generate_or_derive_esk_internal(&mut rng);
-                    let epk = esk.derive_public(
-                        dummy_note
-                            .recipient()
-                            .diversifier()
-                            .g_d()
-                            .expect("checked at construction")
-                            .into(),
-                    );
-
-                    let (zkproof, cv) = prover.output_proof(
-                        ctx,
-                        esk.0,
-                        dummy_note.recipient(),
-                        dummy_note.rcm(),
-                        dummy_note.value().inner(),
-                    );
-
-                    let cmu = dummy_note.cmu();
-
-                    let mut enc_ciphertext = [0u8; 580];
-                    let mut out_ciphertext = [0u8; 80];
-                    rng.fill_bytes(&mut enc_ciphertext[..]);
-                    rng.fill_bytes(&mut out_ciphertext[..]);
-
-                    OutputDescription {
-                        cv,
-                        cmu,
-                        ephemeral_key: epk.to_bytes(),
-                        enc_ciphertext,
-                        out_ciphertext,
-                        zkproof,
-                    }
-                };
-
-                // Update progress and send a notification on the channel
-                progress += 1;
-                if let Some(sender) = progress_notifier {
-                    // If the send fails, we should ignore the error, not crash.
-                    sender
-                        .send(Progress::new(progress, Some(total_progress)))
-                        .unwrap_or(());
+                    SaplingOutputInfo::dummy(&self.params, &mut rng, target_height)
                 }
-
-                result
             })
-            .collect();
+            .collect::<Vec<_>>();
+
+        // Compute the transaction binding signing key.
+        let bsk = {
+            let spends: TrapdoorSum = spend_infos.iter().map(|spend| &spend.rcv).sum();
+            let outputs: TrapdoorSum = output_infos.iter().map(|output| &output.rcv).sum();
+            (spends - outputs).into_bsk()
+        };
+
+        // Create the unauthorized Spend and Output descriptions.
+        let shielded_spends = spend_infos
+            .into_iter()
+            .map(|a| a.build::<SP>(self.anchor))
+            .collect::<Result<Vec<_>, _>>()?;
+        let shielded_outputs = output_infos
+            .into_iter()
+            .map(|a| a.build::<P, OP, _>(&mut rng))
```

Comment:
Okay, so I see that we're propagating the type ascription in to here. I guess that this `build` calls uses methods on `OutputProver` that don't take a `self` reference, and are instead strictly determined by the static type.

Reply:
Correct. This is not necessary for Orchard because it uses a concrete type to hold its proof inputs, whereas here we need to indirectly construct that type via the trait. The aforementioned simplification to not allow remote proving would remove this need.